ja/lessons/basics/functions.md

```diff
@@ -110,15 +119,177 @@ iex> Greeter2.hello("Fred", "Jane")
 "Hello, Fred and Jane"
 ```
 
-関数名の一覧を上記のコメントに載せました。例えば、1つめの実装は引数を取らないので`hello/0`、2つ目は1つの引数を取るので`hello/1`となります。他の言語におけるオーバーロードとは違い、これらは互いに _異なる_ 関数として扱われます。(さっき扱ったパターンマッチングは _同じ_ 数の引数を取る関数定義が複数ある場合のみ適用されます)
+関数名の一覧を上記のコメントに載せました。
+例えば、1つめの実装は引数を取らないので`hello/0`、2つ目は1つの引数を取るので`hello/1`となります。
+他の言語におけるオーバーロードとは違い、これらは互いに _異なる_ 関数として扱われます。
+(さっき扱ったパターンマッチングは _同じ_ 数の引数を取る関数定義が複数ある場合のみ適用されます)
+
+### 関数とパターンマッチング
+
+内部では、関数は実行された時の引数をパターンマッチングしています。
+
+マップを受け取るが、特定のキーにだけ関心がある関数が必要であるとしましょう。
+私たちは次のようにキーの有無に基づいて引数をパターンマッチすることができます:
+
+```elixir
+defmodule Greeter1 do
+  def hello(%{name: person_name}) do
+    IO.puts "Hello, " <> person_name
+  end
+end
+```
+
+今度はFredという名前の人物を表すマップを持っているとしましょう。
+```elixir
+iex> fred = %{
+...> name: "Fred",
+...> age: "95",
+...> favorite_color: "Taupe"
+...> }
+```
+
+`Greeter1.hello/1  `を`fred`のマップで実行するとこのような結果となります:
+
+```elixir
+# call with entire map
+...> Greeter1.hello(fred)
+"Hello, Fred"
+```
+
+`:name`キーを含まないマップで関数を実行するとどうなるでしょうか？
+
+```elixir
+# call without the key we need returns an error
+...> Greeter1.hello(%{age: "95", favorite_color: "Taupe"})
+** (FunctionClauseError) no function clause matching in Greeter3.hello/1
+
+    The following arguments were given to Greeter3.hello/1:
+
+        # 1
+        %{age: "95"}
+
+    iex:12: Greeter3.hello/1
+
+```
+
+このような挙動となる理由は、Elixirは関数が実行された際の引数を関数で定義されたアリティに対してパターンマッチさせているためです。
+
+`Greeter1.hello/1`にデータが届いた際どのように見えるか考えてみましょう:
+
+```Elixir
+# incoming map
+iex> fred = %{
+...> name: "Fred",
+...> age: "95",
+...> favorite_color: "Taupe"
+...> }
+```
+`Greeter1.hello/1`は次のような引数を期待します:
+```elixir
+%{name: person_name}
+```
+`Greeter1.hello/1`では、私たちが渡したマップ(`fred`)は引数(`%{name: person_name}`)に対して評価されます:
+
+```elixir
+%{name: person_name} = %{name: "Fred", age: "95", favorite_color: "Taupe"}
+```
+
+これは渡されたマップの中に`name`に対応するキーを見つけます。
+マッチがありました！このマッチの成功によって、右辺のマップ(つまり
+`fred`マップ)の中にある`:name`キーの値は左辺の変数(`person_name`)に格納されます。 
+
+さて、Fredの名前を`person_name`にアサインしたいが、全体の人物マップの値も保持したいという場合はどうするのでしょう？彼に挨拶した後`IO.inspect(fred)`を使いたいとしましょう。
```

Comment:
I suggest to translate `挨拶を出力した後` than `彼に挨拶した後`.
Who read this article could feel hard to imagine where they put `IO.inspect`, (although It will be explained with code later).
How do you think?

Reply:
Makes sense. I'll fix it!